src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,480 @@
+                                                        count, SIZE_MAX);
+    ssize_t max_short   = ucp_proto_get_short_max(req, msg_config);
+    
+    ucs_status_t status = ucp_request_send_start(req, max_short, 
+                                                 zcopy_thresh, SIZE_MAX,
+                                                 count, msg_config,
+                                                 proto);
+    if (status != UCS_OK) {
+       return UCS_STATUS_PTR(status);
+    }
+
+    /* Start the request.
+     * If it is completed immediately, release the request and return the status.
+     * Otherwise, return the request.
+     */
+    status = ucp_request_send(req);
+    if (req->flags & UCP_REQUEST_FLAG_COMPLETED) {
+        ucs_trace_req("releasing send request %p, returning status %s", req,
+                      ucs_status_string(status));
+        ucp_request_put(req);
+        return UCS_STATUS_PTR(status);
+    }
+
+    ucp_request_set_callback(req, send.cb, cb);
+    ucs_trace_req("returning send request %p", req);
+    
+    return req + 1;
+}
+
+ucs_status_ptr_t ucp_am_send_nb(ucp_ep_h ep, uint16_t id,
+                                const void *payload, size_t count,
+                                uintptr_t datatype, ucp_send_callback_t cb,
+                                unsigned flags)
+{
+    ucs_status_t status;
+    ucs_status_ptr_t ret;
+    size_t length;
+    ucp_request_t *req;
+
+    if (ucs_unlikely(flags != 0)) {
+        ret = UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
+        return ret;
+    }
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    status = ucp_ep_resolve_dest_ep_ptr(ep, ep->am_lane);
+    if (status != UCS_OK) {
+        ret = UCS_STATUS_PTR(status);
+        goto out;
+    }
+    if (ucs_likely(UCP_DT_IS_CONTIG(datatype))) {
+        length = ucp_contig_dt_length(datatype, count);
+        if (ucs_likely((ssize_t)length <= ucp_ep_config(ep)->am.max_short)) {
+            status = ucp_am_send_short(ep, id, payload, length);
+            if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
+                UCP_EP_STAT_TAG_OP(ep, EAGER);
+                ret = UCS_STATUS_PTR(status);
+                goto out;
+            }
+        }
+    }
+    req = ucp_request_get(ep->worker);
+    if (ucs_unlikely(req == NULL)) {
+        ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
+    }
+
+    ucp_am_send_req_init(req, ep, payload, datatype, count, flags, id);
+  
+    ret = ucp_am_send_req(req, count, &ucp_ep_config(ep)->am, cb,
+                          ucp_ep_config(ep)->am_u.proto);
+
+out:
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+    return ret;
+}
+
+static ucs_status_t 
+ucp_am_handler(void *am_arg, void *am_data, size_t am_length,
+               unsigned am_flags)
+{
+    ucp_worker_h worker = (ucp_worker_h) am_arg;
+    ucp_am_hdr_t *hdr = (ucp_am_hdr_t *) am_data;
+    ucs_status_t status;
+    ucp_recv_desc_t *desc = NULL;
```

Comment:
This value is not used, `desc` will be re-initialized in `ucp_recv_desc_init`

Reply:
The compiler does not like the possibility of it not being initialized.